mri/reconstructors/calibrationless.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+This implements the single channel reconstruction.
+"""
+
+from .base import ReconstructorWaveletBase
+from mri.optimizers.utils.cost import GenericCost
+from mri.operators import GradSynthesis, GradAnalysis
+from mri.optimizers import pogm, condatvu, fista
+
+from modopt.opt.proximity import SparseThreshold
+from modopt.opt.linear import Identity
+
+
+class SparseCalibrationlessReconstructor(ReconstructorWaveletBase):
+    """ This class implements a calibrationless reconstruction based on the
+    L1-norm regularization.
+    For the Analysis case finds the solution for x of:
+        (1/2) * sum(||F x_l - yl||^2_2, n_coils) +
+                    mu * sum(||Wt x_l||_1, n_coils)
+    For the Synthesis case finds the solution of:
+        (1/2) * sum(||F Wt alpha_l - yl||^2_2, n_coils) +
+                    mu * sum(||alpha_l||_1, n_coils)
+    Parameters
+    ----------
+    kspace_data: ndarray
```

Comment:
So this is one big question I have with regards to this interface.
I don't understand why you would want to instantiate one reconstructor per image to reconstruct (and even per `kspace_loc`). 
I can speak about my use cases.
Typically, for the fastMRI dataset, I will want to reconstruct several different images with the same config but with different masks (i.e. `kspace_loc`). So I would want to be able to change just the observed data and the mask.

However, when using for example Sparkling trajectories, I will want to keep `kspace_loc` the same, so that the NFFT doesn't have to recompute a plan everytime. Therefore I want to be able to feed in only the images, but different ones at reconstruction time.

Reply:
Please note the lipchitz constant is the limiting factor for us. Should we not have to recalculate lipchitz constant if the kspace_loc changes? I agree that the value is usually in same ball park, but, bad values could result in poor reconstructions (in fact no reconstructions in some cases)

Reply:
well even if you have to it doesn't matter because I would have to do it in the 2 cases: whether I feed it to the reconstruct function or I reinstantiate a Reconstructor object.

Reply:
So your point is to move to a case where we calculate Lipchitz constant at the beginning of reconstruct function? I dont mind that, just that my use case is:
Most of the time with reconstructions, we come up to a case where, lipchitz constant calculation takes too long, and in case of a crash, I would prefer to restart optimization with initialization at the last saved result, and the same lipchitz constant. This would save a lot of overheads and make the codes robust to actual crashes due to memory limitations.

@LElgueddari , can you please shed some light on what your thoughts are on this?

Reply:
It's not really a use case, and I also I didn't quite understand what you were describing.
Also in the api I am advocating, nothing is preventing you from lazily computing the lipschitz constant. That is, you would have an option in the reconstruct function, `compute_lips_cst`, that would allow to compute the lipschitz constant only if it was never computed before if set to `True`.

Reply:
I think that @zaccharieramzi has a good point, one solution would be to remove the k-space values, and to put it into the `reconstruct` function.
However, I suppose that this solution might also impact the way the gradient might be initialized because as far as I remember the observed data are given in the initialization...
But it's true that this is a use-case (that I never met before) which could be solved by adding the observed data as input varaible of the reconstruct function

Reply:
I think we can work on it surely, ie it is doable, the question is whether to do in this PR or another. 
I wouldnt mind doing it here, its @zaccharieramzi  and @LElgueddari your call, as you may have to review it.

Reply:
I think it has to be done here.
@LElgueddari you have never met this use case before? As in you have never wanted to reconstruct several images with the same reconstruction parameters?

It will definitely have an impact on how the gradient is used in the object.

Reply:
I will do it here and re-request the review.

Reply:
I never faced such a use case but I think it's the best thing to do and as you said it has to be done right here,

Reply:
Latest commit fixes this. Resolving this here. Please start a new review for related issues given that we have reached a logical conclusion here.